lightning/src/ln/channelmonitor.rs

```diff
@@ -183,27 +183,32 @@ impl<Key : Send + cmp::Eq + hash::Hash, ChanSigner: ChannelKeys, T: Deref + Sync
 	      L::Target: Logger,
         C::Target: ChainWatchInterface,
 {
-	fn block_connected(&self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], _indexes_of_txn_matched: &[usize]) {
-		let block_hash = header.bitcoin_hash();
-		{
-			let mut monitors = self.monitors.lock().unwrap();
-			for monitor in monitors.values_mut() {
-				let txn_outputs = monitor.block_connected(txn_matched, height, &block_hash, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
-
-				for (ref txid, ref outputs) in txn_outputs {
-					for (idx, output) in outputs.iter().enumerate() {
-						self.chain_monitor.install_watch_outpoint((txid.clone(), idx as u32), &output.script_pubkey);
+	fn block_connected(&self, header: &BlockHeader, txdata: &[(usize, &Transaction)], height: u32) {
+		let mut reentered = true;
+		while reentered {
+			let matched_indexes: HashSet<_> = self.chain_monitor.filter_block(header, txdata).into_iter().collect();
+			let matched_txn: Vec<_> = txdata.iter().cloned().filter(|&(index, _)| matched_indexes.contains(&index)).collect();
+			let last_seen = self.chain_monitor.reentered();
+			{
+				let mut monitors = self.monitors.lock().unwrap();
+				for monitor in monitors.values_mut() {
+					let txn_outputs = monitor.block_connected(header, &matched_txn, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+
+					for (ref txid, ref outputs) in txn_outputs {
+						for (idx, output) in outputs.iter().enumerate() {
+							self.chain_monitor.install_watch_outpoint((txid.clone(), idx as u32), &output.script_pubkey);
+						}
```

Comment:
I like the direction of shifting this logic down into the `SimpleManyChannelMonitor` (SMCM). 

Would the next step towards actually supporting partial blocks be, maybe, returning a list of txs that the SMCM needs more information about? That way, in theory the BlockNotifier, could try to retrieve that additional info from the LN node's block source. If so, or if there's other ideas for better support for partial blocks, could be worth filing an issue.

Reply:
That's a good question. Your suggestion seems like a reasonable approach. I'd like to see if this could be unified somehow with `ChainWatchInterface` by possibly removing it entirely or moving it (or something like it) out to the caller of `BlockNotifier`.

As it stands, the `reentered` logic isn't very intuitive. Ideally it can be handled at the `BlockSource` level such that the caller of `BlockNotifier` can determine if the `BlockSource` needs to be re-queried if filtered blocks were provided. Then perhaps `BlockNotifier` would only ever send filtered blocks to listeners.